enterprise/utils.py

```diff
@@ -601,15 +601,29 @@ def wrapper(request, *args, **kwargs):
 
         enterprise_uuid = kwargs['enterprise_uuid']
         enterprise_customer = get_enterprise_customer_or_404(enterprise_uuid)
+        provider_id = enterprise_customer.identity_provider or ''
+        sso_provider = None
+
+        try:
+            sso_provider = get_identity_provider(provider_id)
+        except ValueError:
+            pass
+
+        decorator_already_processed = request.GET.get('session_cleared') == 'yes'
 
-        # Now verify if the user is logged in. If user is not logged in then
-        # send the user to the login screen to sign in with an
-        # Enterprise-linked SSO and the pipeline will get them back here.
         if not request.user.is_authenticated():
-            next_url = '{current_url}?{query_string}'.format(
+            # If the user isn't logged in, send them to the log in page and redirect them
+            # back to the original view, indicating that the decorator has been processed.
+            next_url = '{current_url}?{params}'.format(
                 current_url=quote(request.get_full_path()),
-                query_string=urlencode({'tpa_hint': enterprise_customer.identity_provider})
+                params=urlencode(
+                    [
+                        ('tpa_hint', provider_id),
+                        ('session_cleared', 'yes')
+                    ]
```

Comment:
actually after changing this, i think the dict makes it so that the order is non deterministic, which is a minor testing headache, so i'm changing it back :-)